### PR TITLE
Clone GitHub repos to the local Gitea during install

### PR DESCRIPTION
**Pull request by kispaljr**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

It clones GitHub repositories that are registered in porch to the local Gitea during install.
This is needed to make the sandbox more self-contained and it is also a first step toward making it work behind a corporate http proxy (because it avoids Internet traffic originating in the clusters).

**Which issue(s) this PR fixes**:

It is a step toward fixing https://github.com/nephio-project/nephio/issues/409 , but it will not completely fix the issue.

** Notes to reviewer:
Git cloning is deliberately done directly on the host machine instead of a pod running in a management cluster. This way we only need to assume that the localhost can reach the Internet (i.e. http proxy vars are setup correctly),  but we don't have to assume that workloads running in the clusters can also reach the internet (that would require the errorprone setting of the no_proxy variable, that is almost impossible to do correctly.)

**Does this PR introduce a user-facing change?**:

no.

### CODE REVIEW

File: e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+- name: "Pull GitHub repo {{ item.name }}"
+  ansible.builtin.git:
+    repo: "{{ item.repo_uri }}"
+    dest: "/tmp/repository/{{ item.name }}.git"
+    bare: yes
+
+- name: "Add gitea remote to local git repo"
+  ansible.builtin.shell: |
+    cd /tmp/repository/{{ item.name }}.git ; git remote add gitea http://nephio:secret@172.18.0.200:3000/nephio/{{ item.name }}.git
```
Comment (electrocucaracha) [inline e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml]: I think that the IP address needs to be discovered

```
- name: Get the gitea IP address
  ansible.builtin.command: kubectl get services -n gitea gitea -o jsonpath='{.status.loadBalancer.ingress[0].ip}'
  register: gitea_ip
```

Comment (electrocucaracha) [inline e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml]:
```suggestion
  ansible.builtin.command:
    cmd: git remote add gitea http://nephio:secret@{{ gitea_ip}}:3000/nephio/{{ item.name }}.git
    chdir: /tmp/repository/{{ item.name }}.git
```

**Comment by kispaljr** [comment]
/assign @electrocucaracha 

@electrocucaracha, I see that you also made changes on how we handle stock git repos. Could you have a look if this was in line with your thinking?

File: e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+- name: "Push tags to gitea: {{ item.name }}"
+  ansible.builtin.shell: |
+    cd /tmp/repository/{{ item.name }}.git ; no_proxy="$no_proxy,172.18.0.200" NO_PROXY="$NO_PROXY,172.18.0.200"  git push gitea --tags --force
```
Comment (electrocucaracha) [inline e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml]:
```suggestion
  environment:
    no_proxy: "{{ gitea_ip}}"
    NO_PROXY: "{{ gitea_ip}}"
  ansible.builtin.command:
    cmd: git push gitea --tags --force
    chdir: /tmp/repository/{{ item.name }}.git
```

File: e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  register: push_result
+  until: push_result is not failed
+  retries: 60
+  delay: 10
+
+- name: "Push branches to gitea: {{ item.name }}"
+  ansible.builtin.shell: |
+    cd /tmp/repository/{{ item.name }}.git ; no_proxy="$no_proxy,172.18.0.200" NO_PROXY="$NO_PROXY,172.18.0.200"  git push gitea --all --force
```
Comment (electrocucaracha) [inline e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml]:
```suggestion
  environment:
    no_proxy: "{{ gitea_ip}}"
    NO_PROXY: "{{ gitea_ip}}"
  ansible.builtin.command:
    cmd: git push gitea --all --force
    chdir: /tmp/repository/{{ item.name }}.git
```

File: e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        labels:
+          kpt.dev/repository-access: read-only
+          kpt.dev/repository-content: external-blueprints
+      spec:
+        content: Package
+        deployment: false
+        git:
+          branch: main
+          directory: /
+          repo: "http://172.18.0.200:3000/nephio/{{ item.name }}.git"
```
Comment (electrocucaracha) [inline e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml]:
```suggestion
          repo: "http://{{ gitea_ip }}:3000/nephio/{{ item.name }}.git"
```

File: e2e/provision/playbooks/cluster.yml [hunk 5]
```diff
@@ -96,6 +96,16 @@
     - bootstrap
     - install
   tasks:
+
+    # Git cloning is deliberately done directly on the localhost instead of a pod running in a management cluster.
+    # This way we only need to assume that the localhost can reach the Internet (i.e. http proxy vars are setup correctly),
```
Comment (electrocucaracha) [inline e2e/provision/playbooks/cluster.yml]:
```suggestion
    # This way we only need to assume that the host machine can reach the Internet (i.e. http proxy vars are setup correctly),
```

**Review by electrocucaracha — COMMENTED** [review]
I'm not able to test it, but I made few suggestions on it. Don't forget to include the license headers on the files

**Comment by electrocucaracha** [comment]
> /assign @electrocucaracha
> 
> @electrocucaracha, I see that you also made changes on how we handle stock git repos. Could you have a look if this was in line with your thinking?

I put some comments on the draft. I think that makes sense, my only concern now it that prow hasn't triggered the CI yet.

File: e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml (hunk 1 above)
Comment (kispaljr) [reply e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml]: right now it is hardcoded in multiple packages in nephio-example-packages.git, but you are right, let's avoid hardcoding it wherever we can.

Comment (kispaljr) [reply e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml]: this is much nicer, thank you

File: e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml (hunk 2 above)
Comment (kispaljr) [reply e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml]: again, much nicer this way. I will change it

File: e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml (hunk 3 above)
Comment (kispaljr) [reply e2e/provision/playbooks/mirror_git_repo_to_gitea.yaml]: same here

File: e2e/provision/playbooks/cluster.yml (hunk 5 above)
Comment (kispaljr) [reply e2e/provision/playbooks/cluster.yml]: this comment was removed along with all changes to cluster.yaml, when I migrated all mirroring into the install role (inspired by your previous commit).

**Comment by kispaljr** [comment]
> > /assign @electrocucaracha
> > @electrocucaracha, I see that you also made changes on how we handle stock git repos. Could you have a look if this was in line with your thinking?
> 
> I put some comments on the draft. I think that makes sense, my only concern now it that prow hasn't triggered the CI yet.

I think somebody (maybe @radoslawc  ?) has to add an ok-to-test label to enable the CI tests

**Comment by kispaljr** [comment]
Can someone help me with the install-integration and provision-linter tests? Is it an error in my code or in the tests?

**Comment by radoslawc** [comment]
> Can someone help me with the install-integration and provision-linter tests? Is it an error in my code or in the tests?

@kispaljr that's a problem with test images, I've pushed the fix here #213

**Comment by kispaljr** [comment]
@electrocucaracha the auto-detection of gitea IP was failing in the install role's molecule test, since there is no deployed gitea in the test environment. I changed the role so that the IP (and the port) is expected to be passed as a fact (~input parameter). Do you think that this is a good solution?

**Comment by kispaljr** [comment]
The ansible linter fails for my PR with these findings:
```yaml[trailing-spaces]: Trailing spaces (warning)
nephio.yaml:20
yaml[line-length]: Line too long (166 > 160 characters) (warning)
nephio.yaml:31
yaml[line-length]: Line too long (180 > 160 characters) (warning)
playbooks/cluster.yml:81
yaml[line-length]: Line too long (181 > 160 characters) (warning)
playbooks/cluster.yml:85
latest[git]: Result of the command may vary on subsequent runs.
playbooks/roles/install/tasks/mirror_git_repo_to_gitea.yaml:24 Task/Handler: Pull GitHub repo {{ item.name }}
command-instead-of-module: git used in place of git module
playbooks/roles/install/tasks/mirror_git_repo_to_gitea.yaml:30 Task/Handler: Add gitea remote to local git repo
command-instead-of-module: git used in place of git module
playbooks/roles/install/tasks/mirror_git_repo_to_gitea.yaml:38 Task/Handler: Push tags to gitea: {{ item.name }}
command-instead-of-module: git used in place of git module
playbooks/roles/install/tasks/mirror_git_repo_to_gitea.yaml:[51](https://prow.nephio.io/spyglass/lens/buildlog/iframe?req=%7B%22artifacts%22%3A%5B%22build-log.txt%22%5D%2C%22index%22%3A1%2C%22src%22%3A%22gs%2Fprow-nephio-sig-release%2Fpr-logs%2Fpull%2Fnephio-project_test-infra%2F212%2Fprovision-linter%2F1721270130563878912%22%7D&topURL=https%3A//prow.nephio.io/view/gs/prow-nephio-sig-release/pr-logs/pull/nephio-project_test-infra/212/provision-linter/1721270130563878912&lensIndex=1#build-log.txt:51) Task/Handler: Push branches to gitea: {{ item.name }}```


- The first 4 of these findings are not in my code. Should I fix them in this PR, anyway?

- The idempotency warning for using git pull: I cannot help it, it is hard to mirror a git repo without it. I could use Gitea's repo mirror functionality instead, but:
  - that would mean Internet traffic after the install script has finished, and that is something I want to avoid
  - that would only work with gitea, while the current solution works with any git repo
-  3 warnings about "git used in place of git module": the git module cannot currently do, what I am trying to achieve with those commands.

Are there any suggestions on how to proceed with these linter findings?

**Comment by radoslawc** [comment]
Hi Istvan, unfortunate thing with this linter is that ansible builtin modules often lack the functionality that shell command provides. I'll look into exempting it for the future (or just generate warning instead of error). For now what I can do is to  
/override provision-linter

**Comment by electrocucaracha** [comment]
> Hi @kispaljr unfortunate thing with this linter is that ansible builtin modules often lack the functionality that shell command provides. I'll look into exempting it for the future (or just generate warning instead of error). For now what I can do is to /override provision-linter

I think that you can disable the lint for a specific line using `noqa` [instruction](https://ansible.readthedocs.io/projects/lint/usage/#muting-warnings-to-avoid-false-positives)

**Comment by electrocucaracha** [comment]
@kispaljr for the [failure](http://prow.nephio.io/view/gs/prow-nephio-sig-release/pr-logs/pull/nephio-project_test-infra/212/install-integration/1721512577541869568#1:build-log.txt%3A4801), installation molecule role is re-using the `prepare.yaml` of kpt role in [local](https://github.com/nephio-project/test-infra/blob/main/e2e/provision/playbooks/roles/install/molecule/default/molecule.yml#L33) and [prow](https://github.com/nephio-project/test-infra/blob/main/e2e/provision/playbooks/roles/install/molecule/ci/molecule.yml#L41) executions. So it's necessary to create its own `prepare.yaml`(because now they diverge) with a task that ensures that gitea package is installed and more likely metallb package will be required as well.

**Comment by electrocucaracha** [comment]
> > @kispaljr for the [failure](http://prow.nephio.io/view/gs/prow-nephio-sig-release/pr-logs/pull/nephio-project_test-infra/212/install-integration/1721512577541869568#1:build-log.txt%3A4801), installation molecule role is re-using the `prepare.yaml` of kpt role in [local](https://github.com/nephio-project/test-infra/blob/main/e2e/provision/playbooks/roles/install/molecule/default/molecule.yml#L33) and [prow](https://github.com/nephio-project/test-infra/blob/main/e2e/provision/playbooks/roles/install/molecule/ci/molecule.yml#L41) executions. So it's necessary to create its own `prepare.yaml`(because now they diverge) with a task that ensures that gitea package is installed and more likely metallb package will be required as well.
> 
> OK, thanks for the input. I was wondering if I should make the git repo cloning optional (only do it if a valid local_git_url variable is set), or should I modify the molecule test to deploy gitea as you suggested. I will chose the later then

If you make it optional, eventually we will have to create another molecule test to cover that scenario.

File: e2e/provision/playbooks/roles/install/molecule/default/tasks/deploy_kpt.yml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  ansible.builtin.command:
+    cmd: "kpt live init --namespace default --context {{ k8s.context }}"
+    chdir: "/tmp/{{ item }}"
+  changed_when: false
+
+- name: "Apply kpt package: {{ item }}"
+  ansible.builtin.command:
+    cmd: "kpt live apply --namespace default --reconcile-timeout 3600 --context {{ k8s.context }}"
+    chdir: "/tmp/{{ item }}"
+  changed_when: false
```
Comment (electrocucaracha) [inline e2e/provision/playbooks/roles/install/molecule/default/tasks/deploy_kpt.yml]: There is a `kpt` Ansible [module ](https://github.com/nephio-project/test-infra/blob/main/e2e/provision/playbooks/library/kpt.py)that you can use if you need and also a `kpt` Ansible [role](https://github.com/nephio-project/test-infra/tree/main/e2e/provision/playbooks/roles/kpt) as well.

**Comment by kispaljr** [comment]
@electrocucaracha could you take a look?
- the install role now works as follows:
  - expects a local_git input parameter with the fetch and push URIs of the Nephio management cluster-local git repo (it is Gitea in our case)
  - it is theoretically Gitea-independent, but right now our Repository resource only supports creating a new repo in Gitea
  - it uses only std git operations to copy external (Github) stock repos to the local git server
  - the molecule test of the role is now deploys gitea and metallb
- kpt CLI calls were replaced with the kpt Ansible module, as you suggested
- all linter findings (including the ones not in my code) were fixed, or  was marked with a noqa tag, as you sugested

**Comment by electrocucaracha** [comment]
@kispaljr I started testing this PR and putting some changes on my [fork](https://github.com/nephio-project/test-infra/compare/main...electrocucaracha:test-infra:github-mirroring), I'll do a PR on your branch once it's ready

**Comment by kispaljr** [comment]
> @kispaljr I started testing this PR and putting some changes on my [fork](https://github.com/nephio-project/test-infra/compare/main...electrocucaracha:test-infra:github-mirroring), I'll do a PR on your branch once it's ready

Thank you. I am struggling to figure out why the webui ReplicaSet times out consistently in the install-integration test.

**Comment by electrocucaracha** [comment]
> > @kispaljr I started testing this PR and putting some changes on my [fork](https://github.com/nephio-project/test-infra/compare/main...electrocucaracha:test-infra:github-mirroring), I'll do a PR on your branch once it's ready
> 
> Thank you. I am struggling to figure out why the webui ReplicaSet times out consistently in the install-integration test.

I've submitted a [PR](https://github.com/nokia/nephio-test-infra/pull/1) in your fork to apply most of the suggestions. It's hard to review given that your fork requires rebase some commits, but if you want to take a look [these are the changes](https://github.com/nephio-project/test-infra/compare/main...electrocucaracha:test-infra:github-mirroring)

**Comment by kispaljr** [comment]
> > > @kispaljr I started testing this PR and putting some changes on my [fork](https://github.com/nephio-project/test-infra/compare/main...electrocucaracha:test-infra:github-mirroring), I'll do a PR on your branch once it's ready
> > 
> > 
> > Thank you. I am struggling to figure out why the webui ReplicaSet times out consistently in the install-integration test.
> 
> I've submitted a [PR](https://github.com/nokia/nephio-test-infra/pull/1) in your fork to apply most of the suggestions. It's hard to review given that your fork requires rebase some commits, but if you want to take a look [these are the changes](https://github.com/nephio-project/test-infra/compare/main...electrocucaracha:test-infra:github-mirroring)

I rebased your changes and pushed to the PR as-is. I have two comments:
- unfortunately it didn't solve my problem with the install-integration test, it is still failing consistently with a webui deployment timeout. do you have an idea how to solve that?
- removing all gitea specific input parameters from the install role and replacing it with generic "local git server" parameters was intentional from my part. I'd like to see the install role to be as independent from the actual kind of the git server that we deploy into the management cluster. I see that you reverted those changes. What is the reason behind this?

**Comment by electrocucaracha** [comment]
@kispaljr  I'm still seeing some differences between these changes and [this other](https://github.com/nokia/nephio-test-infra/pull/1/commits/5a5c3425c91576224429634c11fff63f5581e640)

**Comment by kispaljr** [comment]
> @kispaljr I'm still seeing some differences between these changes and [this other](https://github.com/nokia/nephio-test-infra/pull/1/commits/5a5c3425c91576224429634c11fff63f5581e640)

@electrocucaracha , you are right, they were different. So, I made the tip of my PR (012363532d48e96d71806c76b1fd20b83ac5297d)  identical to the tip of your PR (5a5c3425c91576224429634c11fff63f5581e640). 
Unfortunately the install-integration consistently fails the exact same way as before. Do you have any idea why? Should I just adjust a timeout somewhere in the webui deployment process?

**Comment by electrocucaracha** [comment]
> ``> @kispaljr I'm still seeing some differences between these changes and [this other](https://github.com/nokia/nephio-test-infra/pull/1/commits/5a5c3425c91576224429634c11fff63f5581e640)
> 
> @electrocucaracha , you are right, they were different. So, I made the tip of my PR ([0123635](https://github.com/nephio-project/test-infra/commit/012363532d48e96d71806c76b1fd20b83ac5297d)) identical to the tip of your PR ([5a5c342](https://github.com/nephio-project/test-infra/commit/5a5c3425c91576224429634c11fff63f5581e640)). Unfortunately the install-integration test consistently fails the exact same way as before. Do you have any idea why? Should I just adjust a timeout somewhere in the webui deployment process?
> 
> Here is the status of the nephio-webui ReplicaSet in the install-integration test logs for reference:
> 
> ```
> "status": {
>                 "conditions": [
>                     {
>                         "lastTransitionTime": "2023-12-01T17:14:00Z",
>                         "lastUpdateTime": "2023-12-01T17:14:00Z",
>                         "message": "Deployment does not have minimum availability.",
>                         "reason": "MinimumReplicasUnavailable",
>                         "status": "False",
>                         "type": "Available"
>                     },
>                     {
>                         "lastTransitionTime": "2023-12-01T17:24:01Z",
>                         "lastUpdateTime": "2023-12-01T17:24:01Z",
>                         "message": "ReplicaSet \"nephio-webui-5464c8dc64\" has timed out progressing.",
>                         "reason": "ProgressDeadlineExceeded",
>                         "status": "False",
>                         "type": "Progressing"
>                     }
>                 ],
>                 "observedGeneration": 1,
>                 "replicas": 1,
>                 "unavailableReplicas": 1,
>                 "updatedReplicas": 1
> ```

There is a way to run integration tests locally and review the state of the VM, you can use the following instruction `tox -e install -- --destroy never` on the `e2e/provision`

**Comment by efiacor** [comment]
Hi @kispaljr , is this change still needed/relevant?

**Comment by kispaljr** [comment]
no. it was part of a wider effort to  make the poc work behind an http_l proxy, but I gave up on that since.

**Pull request closed without merging**